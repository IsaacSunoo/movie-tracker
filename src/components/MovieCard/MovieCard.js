import React, { Component } from 'react';
import { connect } from 'react-redux';
import { toggleFavorite } from '../../actions/actions';

class MovieCard extends Component {

    render() {
      const { poster_path, id, handleFavorite, favorites } = this.props;
      const imageSrc = 'http://image.tmdb.org/t/p/w500//' + poster_path;
      const background = { backgroundImage: `url(${imageSrc})` };

      const star = (
        <div className="star">
          {!favorites.includes(id) && <i className="fas fa-star" />}
          {favorites.includes(id) && <i className="fas fa-star active-favorite" />}
        </div>
      );
      return (
        <div className='movie-card' style={background}>
          <div onClick={() => {handleFavorite(id)}}>
            {star}
          </div>
        </div>
    )
  }
}

const mapStateToProps = (state) => ({
  favorites: state.favorites
});

const mapDispatchToProps = (dispatch) => ({
  handleFavorite: (id) => dispatch(toggleFavorite(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieCard);