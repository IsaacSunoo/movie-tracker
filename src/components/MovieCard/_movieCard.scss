.movie-card {
    @include card-design();

    &:hover {
        transform: scale(1.05);
    }

    // grid-template-columns: repeat(3, 1fr);
    // grid-gap: 5px;
    // grid-template-areas: "image title title"
    //                      "image date date"
    //                      "image info info"
    //                      "image info info"
    //                      "image info info";

    // &:link, &:active   {
    //     text-decoration: none;
    // }
}

// .card-image {
//     grid-area: image;
//     justify-content: center;
//     margin: 10px 10px 0 0;
// }

// .movie-title {
//     grid-area: title;
// }

// .release-date {
//     grid-area: date;
// }

// .movie-info {
//     display: flex;
//     justify-content: flex-start;
//     grid-area: info;
// }