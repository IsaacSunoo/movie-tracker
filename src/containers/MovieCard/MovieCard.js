import React, { Component } from 'react';
import { connect } from 'react-redux';
import { toggleFavorite } from '../../actions';
import { Link } from 'react-router-dom';
import { postFetch, deleteFetch } from '../../api';
import PropTypes from 'prop-types';

export class MovieCard extends Component {

  handleFavorite = (movieId) => {
    const { favorites, loginUser } = this.props;
    if (this.isEmpty(loginUser)) {
      console.log('NO USER LOGGED IN')
      alert('Please create account to use this feature')
    } else {
      console.log('login user if block entered');
      if (!favorites.includes(movieId)) {
        this.addFavorite();
      } else {
        this.deleteFavorite(movieId);
      }
    }
  }

  isEmpty(obj) {
    for (var key in obj) {
      if (obj.hasOwnProperty(key))
        return false;
    }
    return true;
  }

  addFavorite = async () => {
    const url = 'users/favorites/new';
    const { id, title, poster_path, release_date, vote_average, overview, loginUser, toggleFavorite } = this.props;
    const favInfo = { movie_id: id, user_id: loginUser.id, title, poster_path, release_date, vote_average, overview };
    await postFetch(url, 'POST', favInfo);
    toggleFavorite(id);
  }

  deleteFavorite = async (movieId) => {
    const { id, toggleFavorite, loginUser } = this.props;
    const url = `users/${loginUser.id}/favorites/${movieId}`;
    await deleteFetch(url);
    toggleFavorite(id);
  }

    render() {
      const { poster_path, id, favorites } = this.props;
      const imageSrc = 'http://image.tmdb.org/t/p/w500//' + poster_path;
      const background = { backgroundImage: `url(${imageSrc})` };

      const star = (
        <div className="star">
          {!favorites.includes(id) && <i className="fas fa-star" />}
          {favorites.includes(id) && <i className="fas fa-star active-favorite" />}
        </div>
      );

      return (
        <div className='movie-card' style={background}>
          <Link to={`/movie/${id}`} key={id}>
            <div className='linkdiv'>
            </div>
          </Link>
          <div onClick={() => {this.handleFavorite(id)}}>
            {star}
          </div>
        </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  loginUser: state.loginUser,
  favorites: state.favorites
});

export const mapDispatchToProps = (dispatch) => ({
  toggleFavorite: (id) => dispatch(toggleFavorite(id))
});

MovieCard.propTypes = {
  loginUser: PropTypes.object.isRequired,
  favorites: PropTypes.array.isRequired,
  toggleFavorite: PropTypes.func.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(MovieCard);