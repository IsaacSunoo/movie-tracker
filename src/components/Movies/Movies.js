import React, { Component } from 'react';
import MovieCard from '../MovieCard/MovieCard';
import { connect } from 'react-redux';
import apiKey from '../../api_key/apiKey';

export class Movies extends Component {
  constructor() {
    super();
    this.state= {
      favoriteMovies: []
    }
  }

  componentDidMount = () => {
    const { favorites } = this.props
    this.getFavorites(favorites)
  }

  fetchMovie = async (url) => {
    let movieUrl = `${url}${apiKey}&language=en-US`;
    const response = await fetch(movieUrl);
    const movies = await response.json();
    return movies;
  }

  getFavorites = async (favorites) => {
    try {
      let unresolvedFavorites = favorites.map(async (movieID) => {
        const url = `https://api.themoviedb.org/3/movie/${movieID}?`
        const favMovie = await this.fetchMovie(url);
        return favMovie;
      })
      let favoriteMovies = await Promise.all(unresolvedFavorites);
      this.setState({ favoriteMovies })
      console.log('fave movies state', this.state.favoriteMovies)
    } catch(err) {
      console.log(err)
    }
  }

  render() {
    const { movies, favorites, id } = this.props;
    const { favoriteMovies } = this.state;
    console.log('favorites: ', favorites);
    console.log('movie props', this.props.id)
    // console.log('props in movie component', movies.results)
    const displayRecentMovies = movies.map(movie => (
        <MovieCard {...movie} key={movie.title} />
    ))

    const displayFavoriteMovies = favoriteMovies.map(movie => (
      <MovieCard {...movie} key={movie.title} />
    ))

    if (id === 'recent') {
      return (
        <div className='movie-container'>
          {displayRecentMovies}
        </div>
      )
    } else if (id === 'favorites') {
      return (
        <div className='movie-container'>
          {displayFavoriteMovies}
        </div>
      )
    }
  }
}

export const mapStateToProps = (state) => ({
  favorites: state.favorites
})

export default connect(mapStateToProps, null)(Movies);