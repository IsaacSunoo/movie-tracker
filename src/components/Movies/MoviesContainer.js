import { connect } from 'react-redux';
import Movies from './Movies';

const mapStateToProps = (state) => ({
    movies: state.movies
});

const mapDispatchToProps = (dispatch) => ({
    getMovies: (movies) => dispatch(getMovies(movies))
});

export default connect(mapStateToProps, mapDispatchToProps)(Movies);