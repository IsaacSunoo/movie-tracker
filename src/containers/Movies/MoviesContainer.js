import { connect } from 'react-redux';
import { getMovies } from '../../actions';
import Movies from './Movies';

const mapStateToProps = (state) => ({
    movies: state.movies
});

const mapDispatchToProps = (dispatch) => ({
    getMovies: (movies) => dispatch(getMovies(movies))
});

export default connect(mapStateToProps, mapDispatchToProps)(Movies);