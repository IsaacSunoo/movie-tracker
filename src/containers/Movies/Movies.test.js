import React from 'react';
import { MovieCard } from '../MovieCard/MovieCard'
import { Movies, mapStateToProps } from './Movies';
import { shallow } from 'enzyme';



describe('Movies', () => {
  let wrapper;
  let mockMovies;
  beforeEach(() => {
    mockMovies = [{title:'hey'}, {title:'what?'}]
    wrapper = shallow(
      <Movies 
        movies= {mockMovies}
      />
      )
    const displayRecentMovies = <MovieCard />;
  });
  
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state', () => {
    expect(wrapper.state()).toEqual({
      favoriteMovies: []
    });
  });
  describe('mapStateToProps', () => {
    it('should return an object with favorites', () => {
      const mockState = {
        loginUser: {name: 'Isaac'},
        favorites: [123, 444],
        movies: [{title: 'Hey'}, {title: 'anotherMovie'}]
      }
      const expected = {
        favorites: [123, 444]
      }
      const result = mapStateToProps(mockState);
      expect(result).toEqual(expected)
    });
  });

  describe('should getFavorites', () => {

  });

  describe('should fetchMovie', () => {

  });
});