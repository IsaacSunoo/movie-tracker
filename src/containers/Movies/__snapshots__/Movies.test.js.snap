// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Movies
    movies={
      Array [
        Object {
          "title": "hey",
        },
        Object {
          "title": "what?",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [
    undefined,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
